### ISSUE TITLE
DPO training produces nan for rewards

### ISSUE THREAD

**Issue by kangqi-ni** [huggingface/trl#878]
I am using DPO trainer to train a vicuna-based fine-tuned model on a custom dataset. However, after a few steps, the trainer started to output nan for rewards. Did everyone run into similar issues or know how to fix this? 

Edit: I tried using different models or use a public dataset "lvwerra/stack-exchange-paired". I still ran into nan values starting on the second step.

![Screenshot from 2023-10-15 19-48-23](https://github.com/huggingface/trl/assets/64222208/863e4252-b488-4128-b8cb-8f935a39f681)

**Comment by lvwerra** [comment]
cc @kashif

**Comment by alex-ht** [comment]
I have the same problem with fp16, but not with bf16.

**Comment by fe1ixxu** [comment]
The same issue with fp16

**Comment by worden1** [comment]
Same issue with fp16 on V100, tried fp32 and load reference model with float32 then seems the nan fixed.

**Comment by younesbelkada** [comment]
I think the fix is to upcast the logits in fp32 to avoid underflow & overflow here: https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L405-L408 Since the rewards are never used for backprop it should be totally fine. Just made: https://github.com/huggingface/trl/pull/942

**Comment by alex-ht** [comment]
> I think the fix is to upcast the logits in fp32 to avoid underflow & overflow here: https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L405-L408 Since the rewards are never used for backprop it should be totally fine. Just made: #942

Should we convert logits to fp32 before applying logsoftmax? Megatron-LM converts it before calculating cross entropy.
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py#L36
https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L437

**Comment by alex-ht** [comment]
All logps are expected to be in torch.FloatTensor, so why upcast them to fp32 again?

**Comment by kashif** [comment]
I have removed the line that cast the `logits` to `float32` unnecessarily and have tested the `dpo.py` script in both `float16` and `bfloat16` mode so I believe that should fix the issue... @kangqi-ni can you tell me how to reproduce your error exactly?

**Comment by OmarMohammed88** [comment]
@younesbelkada  i tried your solution but it still produce nan values after set of steps

**Comment by younesbelkada** [comment]
Thanks! Can you share a small reproducible snippet?

**Comment by kashif** [comment]
@OmarMohammed88 also note that one typically uses a very small learning rate (something like `1e-7`) when doing DPO training
